llvm/lib/Target/AIE/ReservedRegsLICM.cpp

```diff
@@ -0,0 +1,387 @@
+  void processForExitSink(MachineLoop &L, const BitVector &ReservedLiveins);
+
+  /// Sink \p Cand to \p L's exit block if it is safe to do so.
+  bool trySinkToExitBlock(const CandidateInfo &Cand, MachineLoop &L);
+
+  /// Find instructions to hoist to the preheader
+  void processForPreheaderHoist(MachineLoop &L,
+                                const BitVector &ReservedLiveins);
+
+  /// Hoist \p Cand to \p L's preheader if it is safe to do so.
+  bool tryHoistToPreHeader(const CandidateInfo &Cand, MachineLoop &L);
+
+  /// Verify if \p Cand is loop invariant and can be safely hoisted.
+  /// \pre Cand->DefinedReg has a unique live value within the loop. This is
+  /// verified by processForExitSink() or processForPreheaderHoist().
+  bool isLoopInvariantInst(const CandidateInfo &Cand, const MachineLoop &L);
+};
+
+} // end anonymous namespace
+
+char ReservedRegsLICM::ID;
+
+char &llvm::ReservedRegsLICMID = ReservedRegsLICM::ID;
+
+INITIALIZE_PASS_BEGIN(ReservedRegsLICM, DEBUG_TYPE,
+                      "Machine LICM for reserved regs", false, false)
+INITIALIZE_PASS_DEPENDENCY(MachineLoopInfo)
+INITIALIZE_PASS_END(ReservedRegsLICM, DEBUG_TYPE,
+                    "Machine LICM for reserved regs", false, false)
+
+llvm::FunctionPass *llvm::createReservedRegsLICMPass() {
+  return new ReservedRegsLICM();
+}
+
+bool ReservedRegsLICM::runOnMachineFunction(MachineFunction &MF) {
+  if (skipFunction(MF.getFunction()))
+    return false;
+
+  const TargetSubtargetInfo &ST = MF.getSubtarget();
+  TRI = ST.getRegisterInfo();
+  MRI = &MF.getRegInfo();
+  assert(MRI->isSSA() && "ReservedRegsLICM can only deal with SSA vregs");
+
+  LLVM_DEBUG(dbgs() << "******** Reserved register LICM: " << MF.getName()
+                    << " ********\n");
+
+  MachineLoopInfo &MLI = getAnalysis<MachineLoopInfo>();
+
+  SmallVector<MachineLoop *, 4> Loops = MLI.getBase().getLoopsInPreorder();
+  for (MachineLoop *L : reverse(Loops)) {
+    runOnLoop(*L);
+  }
+
+  return Changed;
+}
+
+BitVector ReservedRegsLICM::collectLoopReservedLiveins(const MachineLoop &L) {
+  const MachineBasicBlock &Header = *L.getHeader();
+  LivePhysRegs LiveRegs(*TRI);
+
+  // Conservativaly assume reserved reserved regs are all liveouts
+  for (MCPhysReg PhysReg : MRI->getReservedRegs().set_bits()) {
+    if (MRI->canSimplifyPhysReg(PhysReg)) {
+      LiveRegs.addReg(PhysReg);
+    }
+  }
+
+  // Traverse instructions to remove defs
+  for (const MachineInstr &MI : reverse(Header))
+    LiveRegs.stepBackward(MI);
+
+  BitVector ReservedLiveins(TRI->getNumRegs());
+  for (MCRegister Reg : LiveRegs) {
+    if (MRI->isReserved(Reg)) {
+      ReservedLiveins.set(Reg);
+    }
+  }
+  return ReservedLiveins;
+}
+
+/// Walk the specified region of the CFG and hoist loop invariants out to the
+/// preheader.
+void ReservedRegsLICM::runOnLoop(MachineLoop &L) {
+  const MachineBasicBlock *Preheader = L.getLoopPreheader();
+  const MachineBasicBlock *ExitBlock = L.getExitBlock();
```

Comment:
Shouldn't you check that it only has one predecessor?

Reply:
Definitely... 🙈 I have a local fix that will create a dedicated exit when required. Checking QoR now.